### PR TITLE
Reintrospection Fix for Many To Many relations

### PR DESCRIPTION
**Pull request by do4gr**
In order to uniquely identify ManyToMany relations we need to take the relation name into account. 

Fixes https://github.com/prisma/prisma/issues/3507
Fixes https://github.com/prisma/prisma/issues/3516

### CODE REVIEW

File: introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs [hunk 1]
```diff
@@ -135,61 +137,40 @@ pub fn enrich(old_data_model: &Datamodel, new_data_model: &mut Datamodel) -> Vec
...
-                relation_fields_to_change.push((changed_model.name.clone(), rf.name.clone(), relation_name.clone()));
-                relation_fields_to_change.push((
-                    other_model_in_relation.name.clone(),
-                    other_relation_field.name.clone(),
-                    relation_name.clone(),
-                ));
+        for model in new_data_model.models() {
+            for field in model.relation_fields() {
+                if let Some(old_model) = old_data_model.find_model(&model.name) {
+                    for old_field in old_model.relation_fields() {
```
Comment (tomhoule) [inline introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs:148]: These nested loops and conditionals are making me think we could structure reintrospection a bit more like the sql schema differ in the migration engine. For example here, there are 4 level of nestings for an action that is conceptually pretty simple: comparing the relation fields for models that are present in both the new and the old schemas. Maybe we could build to an API like `reintrospector.model_pairs().flat_map(|models| models.relation_field_pairs()).for_each(|relation_fields| { /* logic goes here */ })` — in the future.

Comment (tomhoule) [reply introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs:148]: (we can even add a `.filter_map(|relation_fields| relation_fields.as_m2m())` and then be sure we are dealing only with many-to-many relations)

**Review by tomhoule — APPROVED** [review]
The logic looks sound to me :+1:

File: introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs (hunk 1 above)
Comment (do4gr) [reply introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs:148]: Yeah, let's pair on that once we have a bit of time.

**Pull request merged**